cocotb/scheduler.py

```diff
@@ -578,9 +578,23 @@ def queue_function(self, coroutine):
         # each entry always has a unique thread.
         t, = matching_threads
 
+        @coroutine
+        def wrapper():
+            coro = cocotb.coroutine(undecorated_coro)(*args, **kwargs)
+            try:
+                _outcome = outcomes.Value((yield coro))
+            except BaseException as e:
+                _outcome = outcomes.Error(e)
+            event.outcome = _outcome
+            event.set()
+
+        event = threading.Event()
         t.thread_suspend()
-        self._pending_coros.append(coroutine)
-        return t
+        self._pending_coros.append(wrapper())
+        # This blocks the calling external thread until the coroutine finishes
+        event.wait()
+        t.thread_resume()
+        return event.outcome.get()
```

Comment:
Part B (dest) - note that this is the src, with `queue_function` replaced by the lines on the left